tests/contracts/inheritance.algo.ts

```diff
@@ -0,0 +1,41 @@
+/* eslint-disable max-classes-per-file */
+import { Contract } from '../../src/lib/index';
+
+export class A extends Contract {
+  aKey = GlobalStateKey<number>({ key: 'a' });
+
+  a(): void {
+    this.aKey.value = 1337;
+  }
+}
+
+export class B extends A {
+  bKey = GlobalStateKey<number>({ key: 'b' });
+
+  b(): void {
+    assert(this.aKey.value === 1337);
+    this.bKey.value = 42;
+  }
+}
+
+export class C extends B {
+  c(): void {
+    assert(this.bKey.value === 42);
+    assert(this.aKey.value === 1337);
+  }
+}
+
+export class D extends Contract {
+  dKey = GlobalStateKey<number>({ key: 'd' });
+
+  d(): void {
+    this.dKey.value = 42;
+  }
+}
+
+export class AD extends Contract.extend(A, D) {
+  ad(): void {
+    assert(this.aKey.value === 1337);
+    assert(this.dKey.value === 42);
+  }
+}
```

Comment:
So I've just been having a little play and hit a stop.

If I want to call a public or private function from another class, would you expect to use `this.funcName()`? Right now I seem to get an `Unsupported CallExpression` error during compilation, so I guess it's not known about.

Using this section as an example... I can see that `d()` is a callable method from the `D` contract, but how would you call `d()` from inside `ad()` ? And would it being public or private make any difference?

Reply:
Strange... are you on the latest commit? I can't seem to reproduce. The following code compiles fine for me. 

```ts
export class AD extends Contract.extend(A, D) {
  ad(): void {
    this.d()
    assert(this.aKey.value === 1337);
    assert(this.dKey.value === 42);
  }
}
```

As far as public vs private goes, I need to add support for `protected`. This is basically a private method that is also callable by subclasses. Added this to the TODO list

Reply:
Can confirm this was me being stupid. I'd created a bunch of different classes when testing, and somehow forgot which one I was extending. That's why it couldn't find the method.